### PR TITLE
Refactor more structure-based approach to explore/stream

### PR DESCRIPTION
**Pull request by pinnouse**
This pull request factors out common pretty-printing from explore/stream and adds a new structure called explore-stream which is used in explore-state/stream, replicating the same behaviour, but using the structure-first approach and offloading printing to separate functions.

### CODE REVIEW

File: tools.rkt [hunk 1]
```diff
@@ -305,6 +315,87 @@
...
+     (pp/results results qvars)
+     (when (and (previous-choice undo) (null? results))
+       (printf "Previous Choice:\n")
+       (print-choice qvars (previous-choice undo))
+       (newline))
+     (printf "Current Depth: ~a\n" (length undo))
+     (pp/choices choices qvars)
+     ]))
+
+(define (explore-state/stream step qvars strm)
```
Comment (gregr) [inline tools.rkt:369]: We also want to add another parameter for the interaction policy that determines how interaction I/O works.

For organizing your experiments, you really want to maintain multiple policies to see what impact each of them has on the agent's performance.  Being able to swap policies easily is important.

We can also redefine `explore/stream` later, having it call your new driver with a textual pretty-print-and-read interaction policy that mimics the current behavior.

File: tools.rkt [hunk 2]
```diff
@@ -305,6 +315,87 @@
...
+      (define i (read))
+      (cond
+        [(eof-object? i) (newline)]
+        [(or (eq? i 'h) (eq? i 'help))
+         (displayln
+            (string-append "\nType either the letter 'u' or the"
+                           " number following one of the listed choices."
+                           "\nHit enter to continue."))
+         (read-line)
+         (read-line)
```
Comment (gregr) [inline tools.rkt:390]: Keeping in mind the separation described earlier, between updating the state and presentation + receiving of input, this driver should not include any specific strings or printing, and should not use `read` directly. All of those decisions should come from the policy parameter.

File: tools.rkt (hunk 1 above)
Comment (gregr) [inline tools.rkt:369]: Also, this isn't operating on an `explore-state` input, so the name isn't appropriate.  Maybe choose something like `drive/stream` instead.

**Review by gregr — CHANGES_REQUESTED** [review]
The new `explore-state` actuation (`explore-state/make-choice` and `explore-state/make-choices`) and driver loop (`explore-state/stream`) look fine so far, though see the inline comments about naming conventions.

The only real issue is that the pretty-printing logic should not be globally shared like this.  What we want is a driver loop that takes its pretty-printing and user-input policy as a parameter.  More detail in the comments.

**Review by gregr — CHANGES_REQUESTED** [review]
Thanks for the update.

I think there's some confusion about the purpose of separating the driver and policies.  The driver should be carrying out all manipulation of the exploration state.  The policies should define a user interface for displaying the state and receiving commands from the user, then sending those commands to the driver for execution.

This separation is intended to reduce the footprint of future changes while carrying out experiments.  We should be able to swap UIs without changing the exploration code, and we we should be able to adjust the exploration code without large effects on the UI.

It looks like your zipper implementation is on the right track, but could potentially be simplified.  There are more parts than I'd expect to see for a zipper.  If the `explore-loc` is necessary on top of both an `explore-node` and an `explore-context`, an explanation would help.

Please also remove problematic comments and fix indentation.

File: tools.rkt [hunk 3]
```diff
@@ -305,6 +317,112 @@
...
+(struct explore-context (index siblings choices context) #:prefab)
+(struct explore-node (index choices expanded-choices) #:prefab)
+;;       a
+;;    /--|--\
+;;   b   c   d
+;;  / \
+;; e   @
+;; would be modeled as (@ as current context):
+;; (1 '(e) (0 '(c d) (-1 '(a) 'top)))
+(struct explore-loc (tree context) #:prefab)
```
[the remark this answers is not in the document]
Comment (pinnouse) [reply tools.rkt:333]: I found according to what I read from Wikipedia on zippers and the 2 resources on a tree in Haskell:
- https://wiki.haskell.org/Zipper
- https://learnyouahaskell.com/zippers
the zipper technique stores the actual object that is in focus, and the context that the focused object comes from

[the remark this answers is not in the document]
Comment (gregr) [reply tools.rkt:333]: Right, that's two things.  I see you have these three things:
- `explore-context`
- `explore-node`
- `explore-loc`

What role does each one play in the zipper?

[the remark this answers is not in the document]
Comment (pinnouse) [reply tools.rkt:333]: The "two" you would use for zipper are the explore-node (tree in focus), and explore-context (tree with hole). The explore-loc is a struct that just allows the two to be transported together, it's essentially an ordered pair: tree pointing to the node, context pointing to the context.

[the remark this answers is not in the document]
Comment (gregr) [reply tools.rkt:327]: Thanks, that makes sense now.

File: tools.rkt [hunk 4]
```diff
@@ -305,6 +311,131 @@
...
+(define (drive/policy step qvars policy-print policy-read policy-done? init-state)
+  (let loop ([s init-state])
+    (policy-print s qvars)
+    (unless (policy-done? s)
+      (let* ([input (policy-read)]
+             [tree (explore-loc-tree s)])
+        (loop
+          (cond
+            [(and (integer? input) (<= 1 input) (<= input (length (explore-node-choices tree))))
+            (explore-choice s step (- input 1))]
```
Comment (gregr) [inline tools.rkt:424]: Thanks, this looks good now.  Later, if we expand the actions an agent may take, we may have to give more structure to the commands coming from the policy.  But looking for `integer?` vs. `u` is fine for now.

The indentation should have another space.

**Review by gregr — CHANGES_REQUESTED** [review]
Thanks, almost there.  Just a few small issues to fix in naming, indentation, and reverting `explore/stream`.

File: tools.rkt [hunk 5]
```diff
@@ -305,6 +311,131 @@
                  (cons (cons i s) undo)))
           (else (invalid)))))
 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Define explore state to maintain the data held at every step
+;; when exploring
+;; https://wiki.haskell.org/Zipper
+(struct explore-parent-context (index siblings choices context) #:prefab)
```
[the remark this answers is not in the document]
Comment (pinnouse) [reply tools.rkt:318]: That's correct it does refer to its parent context, I am afraid of getting too verbose with the naming; accessing the parent context might look like explore-context-parent-context

[the remark this answers is not in the document]
Comment (gregr) [reply tools.rkt:318]: I think `parent` is still better than `context`, and it would only lead to the accessor name `explore-context-parent`.

**Review by gregr — APPROVED** [review]
Looks good.  Thanks for all the iterating.

**Pull request merged**